ledger/ledger.go

```diff
@@ -187,11 +188,33 @@ func (k *Key) Size() int {
 // Warning, Changing this has an impact on how leaf hashes are computed
 // don't use this to reconstruct the key later
 func (k *Key) CanonicalForm() []byte {
-	ret := ""
+	// calculate the size of the byte array
+
+	// the maximum size of a uint16 is 5 characters, so
+	// this is using 10 for the estimate, to include the two '/'
+	// characters and an extra 3 characters for padding safety
+
+	constant := 10
+
+	requiredLen := constant * len(k.KeyParts)
 	for _, kp := range k.KeyParts {
-		ret += fmt.Sprintf("/%d/%v", kp.Type, string(kp.Value))
+		requiredLen += len(kp.Value)
 	}
-	return []byte(ret)
+
+	retval := make([]byte, 0, requiredLen)
```

Comment:
I tried removing the length estimation part and replaced it with just `var retval []byte`, and I noticed the code became much less performant (time and memory) 😮 
```
BenchmarkCan1-8         13613204                86.0 ns/op            64 B/op          1 allocs/op
BenchmarkCan2-8          6567470               183 ns/op             120 B/op          4 allocs/op
```